syntax = "proto3";

package dev_observer.api.types.changes;

import "google/protobuf/timestamp.proto";

message GitHubChangesSummary {
  string id = 1;
  string repo_id = 2;
  string repo_full_name = 3;
  google.protobuf.Timestamp created_at = 4;
  google.protobuf.Timestamp analysis_period_start = 5;
  google.protobuf.Timestamp analysis_period_end = 6;
  ChangesSummaryContent content = 7;
  ChangesSummaryStatus status = 8;
  optional string error_message = 9;
}

message ChangesSummaryContent {
  string summary = 1;
  repeated CommitInfo commits = 2;
  repeated FileChange file_changes = 3;
  repeated PullRequestInfo pull_requests = 4;
  repeated IssueInfo issues = 5;
  ChangesStatistics statistics = 6;
}

message CommitInfo {
  string sha = 1;
  string message = 2;
  string author = 3;
  google.protobuf.Timestamp committed_at = 4;
  repeated string files_changed = 5;
  int32 additions = 6;
  int32 deletions = 7;
}

message FileChange {
  string file_path = 1;
  ChangeType change_type = 2;
  int32 additions = 3;
  int32 deletions = 4;
  optional string language = 5;
}

message PullRequestInfo {
  int32 number = 1;
  string title = 2;
  string state = 3;
  string author = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp updated_at = 6;
  google.protobuf.Timestamp merged_at = 7;
  repeated string labels = 8;
  int32 additions = 9;
  int32 deletions = 10;
}

message IssueInfo {
  int32 number = 1;
  string title = 2;
  string state = 3;
  string author = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp updated_at = 6;
  google.protobuf.Timestamp closed_at = 7;
  repeated string labels = 8;
}

message ChangesStatistics {
  int32 total_commits = 1;
  int32 total_additions = 2;
  int32 total_deletions = 3;
  int32 total_files_changed = 4;
  int32 total_pull_requests = 5;
  int32 total_issues = 6;
  repeated LanguageStats language_stats = 7;
}

message LanguageStats {
  string language = 1;
  int32 files_changed = 2;
  int32 additions = 3;
  int32 deletions = 4;
}

enum ChangeType {
  CHANGE_TYPE_UNSPECIFIED = 0;
  CHANGE_TYPE_ADDED = 1;
  CHANGE_TYPE_MODIFIED = 2;
  CHANGE_TYPE_DELETED = 3;
  CHANGE_TYPE_RENAMED = 4;
}

enum ChangesSummaryStatus {
  CHANGES_SUMMARY_STATUS_UNSPECIFIED = 0;
  CHANGES_SUMMARY_STATUS_PENDING = 1;
  CHANGES_SUMMARY_STATUS_PROCESSING = 2;
  CHANGES_SUMMARY_STATUS_COMPLETED = 3;
  CHANGES_SUMMARY_STATUS_FAILED = 4;
} 